src/input/json_to_storage.hh

```diff
@@ -123,27 +150,213 @@ public:
     string str();
 
     /**
-     * Put actual address to previous_references_ set
+     * Check if index of Abstract descendant can be loaded from TYPE key (return true)
+     * or return false if automatic conversion can be performed. In other cases throw
+     * input exception.
      */
-    void put_address();
+    virtual bool has_descendent_index(bool value_at_declaration) =0;
+
+    /// Returns name of input format (e. g. "JSON")
+    virtual std::string input_format_name() const =0;
+
+    /// Returns name of sequence type (e. g. "JSON array")
+    virtual std::string sequence_name() const =0;
+
+    /// Returns name of map type (e. g. "JSON object")
+    virtual std::string map_name() const =0;
+
+protected:
+    PathBase();
+
+    std::set<string> previous_references_;
 
-private:
     /**
      * One level of the @p path_ is either index (nonnegative int) in array or string key in a json object.
      * For the first type we save index into first part of the pair and empty string to the second.
      * For the later type of level, we save -1 for index and the key into the secodn part of the pair.
      */
     vector< pair<int, string> > path_;
+
+};
+
+
+
+/**
+ * @brief Class used by JSONToStorage class to iterate over the JSON tree provided by json_spirit library.
+ *
+ * This class keeps whole path from the root of the JSON tree to the current node. We store nodes along path in \p nodes_
+ * and address of the node in \p path_.
+ *
+ * The class also contains methods for processing of special keys 'REF' and 'TYPE'. The reference is record with only one key
+ * 'REF' with a string value that contains address of the reference. The string with the address is extracted by \p JSONToStorage::get_ref_from_head
+ * then the PathJSON corresponding to the address is provided by method \p JSONtoStorage::find_ref_node.
+ */
+class PathJSON : public PathBase {
+public:
+    typedef json_spirit::mValue Node;
+
+    PathJSON(istream &in);
+
+    /**
+     * Dive into json_spirit hierarchy. Store current path and returns true if pointer to new json_spirit node is not NULL.
+     */
+    virtual bool down(unsigned int index) override;
+    virtual bool down(const string& key) override;
+
+    /**
+     * Return one level up in the hierarchy.
+     */
+    virtual void up() override;
+
+    /**
+     * Move to root node.
+     */
+    virtual void go_to_root() override;
+
+    /**
+     * Returns level of actual path. Root has level == 0.
+     */
+    virtual inline int level() const
+    { return nodes_.size() - 1; }
+
+    /**
+     * Check if current head node is a JSON Object containing one key REF of type string.
+     * If yes, returns the string through reference @p ref_address.
+     */
+    bool get_ref_from_head(string & ref_address) override;
+
+    // These methods are derived from PathBase
+    virtual bool is_null_type() const override;
```

Comment:
Use just override no virtual in child classes. This has sense only if you want to allow PathJSON to be a parent of some other class.